### PR TITLE
Shape Selection And Move

### PR DESCRIPTION
**Pull request by MandaBrown**
Connects to #246 

core-demo-select

This PR adds the ability to select and move a shape (no ReactGUI for it yet, so I've only added it to the core demo page). 

Some things: 
  - This is not currently an undoable action. It would be good to make this an action in the future, but not for every step of the way, just start and end.
  - I'm currently using the bounding rectangle as the 'select area' for each shape. I like this because for the whiteboard application that I'm working on we want something simple and this way you don't have to get fiddly with little bits to select something and move it out of the way. An earlier implementation was closer to shape outlines, so if you had transparent shapes you could select what's underneath them by clicking the visible bits, etc, by having each shape draw itself exactly on the select canvas (but in the proper color). I don't know which way we want the main library to go though, so let me know if the bounding rectangle select area isn't ok for now :)

### CODE REVIEW

**Comment by MandaBrown** [comment]
(getting rid of the merge commit)

**Comment by christianheimann** [comment]
It's looking good! We would be glad to have this feature soon =)

**Comment by irskep** [comment]
The behavior you describe is a great start and I would have no problem merging your changes (once I've looked over the code tomorrow).

**Comment by MandaBrown** [comment]
Thanks!

**Comment by irskep** [comment]
Doesn't work for me. Somehow I clicked a selection box for a shape that isn't there:

screenshot 2015-12-18 11 07 35

Here's the JSON of the drawing: https://gist.github.com/irskep/8db1b5715be7e1c33932

**Comment by irskep** [comment]
Oh, I just realized: I was selecting the eraser!

Selecting the polygons didn't seem to work though.

**Comment by irskep** [comment]
I also couldn't select the red line, or the line with the arrow on the end.

File: src/core/LiterallyCanvas.coffee [hunk 1]
```diff
@@ -273,8 +281,12 @@ module.exports = class LiterallyCanvas
...
             @transformed (=>
               for shape in @_shapesInProgress
-                renderShapeToContext(
-                  @ctx, shape, {@bufferCtx, shouldOnlyDrawLatest: true})
+                shape.drawLatest(@ctx, @bufferCtx)
+                if (@selectedShape? && shape.id == @selectedShape.id)
+                  createShape('SelectionBox', {
+                    shape: shape,
+                    handleSize: 0
+                  }).drawLatest(@ctx, @bufferCtx)
```
Comment (irskep) [inline src/core/LiterallyCanvas.coffee]: Even though I think I understand why you did this (so the selection box would be at the same layer as the shape), I think I would prefer it if the selection tool handled this and just drew the box on top of everything by setting the selection box as its "shape in progress."

File: src/core/LiterallyCanvas.coffee [hunk 2]
```diff
@@ -125,6 +124,15 @@ module.exports = class LiterallyCanvas
...
   setShapesInProgress: (newVal) -> @_shapesInProgress = newVal
 
+  selectShape: (shape) ->
+    @selectedShape = shape
+    if @selectedShape?
+      @setShapesInProgress [@selectedShape]
+      @trigger('shapeSelected', { @selectedShape })
+    else
+      @setShapesInProgress []
+    @repaintLayer('main')
```
Comment (irskep) [inline src/core/LiterallyCanvas.coffee]: I think this should move to the tool. Would rather keep this class as small as possible (it's already too large).

File: src/core/LiterallyCanvas.coffee [hunk 3]
```diff
@@ -299,13 +311,8 @@ module.exports = class LiterallyCanvas
...
   # The context is restored to its original state before returning.
   drawShapeInProgress: (shape) ->
+    @setShapesInProgress [shape]
     @repaintLayer('main', false)
-    @clipped (=>
-      @transformed (=>
-        renderShapeToContext(
-          @ctx, shape, {@bufferCtx, shouldOnlyDrawLatest: true})
-      ), @ctx, @bufferCtx
-    ), @ctx, @bufferCtx
```
Comment (irskep) [inline src/core/LiterallyCanvas.coffee]: :+1:

File: src/core/LiterallyCanvas.coffee (hunk 2 above)
Comment (MandaBrown) [inline src/core/LiterallyCanvas.coffee]: Yeah, I see your point! I'll move the selection / selection box drawing to the tool.

**Comment by MandaBrown** [comment]
Yeah, selecting with the eraser as a possible tool is weird (as you mentioned in the issue). :/ Not sure what to do about it? The app that I have that's using this doesn't have an eraser, so it's not an issue there... 

I'll try to figure out what's going on with your shapes and why some of them aren't selecting correctly...

**Comment by irskep** [comment]
No problem about the eraser. I think I said before, that ship sailed with SVG support. If people want erasers they have to give up some features. (In the future, I might implement a raster-based selection box tool that makes images instead of moving shapes, but that's out of scope for you.0

File: src/core/shapes.coffee [hunk 4]
```diff
@@ -132,6 +132,9 @@ defineShape 'Image',
       img = new Image()
       img.src = data.imageSrc
     createShape('Image', {x: data.x, y: data.y, image: img, scale: data.scale})
+  move: ( moveInfo={} ) ->
+    @x = @x - moveInfo.xDiff
+    @y = @y - moveInfo.yDiff
```
Comment (irskep) [inline src/core/shapes.coffee]: I think I would prefer an absolute `setUpperLeft()` instead of delta-based `move()`. As we make more use of it, the logic will be simpler for the caller.

File: src/index.coffee [hunk 5]
```diff
@@ -40,6 +40,7 @@ tools =
   Polygon: require './tools/Polygon'
   Pan: require './tools/Pan'
   Eyedropper: require './tools/Eyedropper'
+  ShapeSelect: require './tools/ShapeSelect'
```
Comment (irskep) [inline src/index.coffee]: This is pretty nitpicky, but I'd switch the order of this name: `SelectShape`. That way if/when I add `SelectBitmap` they alphabetize together.

File: src/tools/ShapeSelect.coffee [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  name: 'ShapeSelect'
+  usesSimpleAPI: false
+
+  constructor: (lc) ->
+    # This is a 'shadow' canvas -- we'll reproduce the shapes here, each shape
+    # with a different color that corresponds to their index. That way we'll
+    # be able to find which shape to select on the main canvas by pixel color.
+    @selectCanvas = document.createElement('canvas')
+    @selectCanvas.style['background-color'] = 'transparent'
+    @selectCtx = @selectCanvas.getContext('2d')
```
Comment (irskep) [inline src/tools/ShapeSelect.coffee]: Is this still used? I thought you said you were doing it based on bounding box?

If you want to make it configurable, that's fine too.

File: src/core/LiterallyCanvas.coffee [hunk 7]
```diff
@@ -313,7 +320,7 @@ module.exports = class LiterallyCanvas
     return unless shapes.length
     drawShapes = =>
       for shape in shapes
-        renderShapeToContext(ctx, shape, {retryCallback})
+        shape.draw(ctx, { retryCallback })
```
Comment (irskep) [inline src/core/LiterallyCanvas.coffee]: Why this change? `shape.draw()`

File: src/tools/ShapeSelect.coffee (hunk 6 above)
Comment (MandaBrown) [inline src/tools/ShapeSelect.coffee]: Yes - the 'bounding box' shapes are drawn onto this canvas...

Comment (irskep) [inline src/tools/ShapeSelect.coffee]: Oh, I get it, sorry. I didn't actually read the whole class.

File: src/tools/ShapeSelect.coffee [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  _drawSelectCanvas: (lc) ->
+    @selectCanvas.width = lc.canvas.width
+    @selectCanvas.height = lc.canvas.height
+    @selectCtx.clearRect(0, 0, @selectCanvas.width, @selectCanvas.height)
+    shapes = lc.shapes.map (shape, index) =>
+      createShape('SelectionBox', {
+        shape: shape,
+        handleSize: 0,
+        backgroundColor: "##{@_intToHex(index)}"
+      })
```
Comment (irskep) [inline src/tools/ShapeSelect.coffee]: So in theory, this could also return a copy of the shape if you wanted to select by pixels rather than bounding box? (Don't need to make it configurable now, but it's good to know.)

File: src/index.coffee (hunk 5 above)
Comment (MandaBrown) [inline src/index.coffee]: Makes sense to me!

File: src/core/shapes.coffee (hunk 4 above)
Comment (MandaBrown) [inline src/core/shapes.coffee]: I could see that. I'll play with doing it that way instead.

File: src/tools/ShapeSelect.coffee [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+  constructor: (lc) ->
+    # This is a 'shadow' canvas -- we'll reproduce the shapes here, each shape
+    # with a different color that corresponds to their index. That way we'll
+    # be able to find which shape to select on the main canvas by pixel color.
+    @selectCanvas = document.createElement('canvas')
+    @selectCanvas.style['background-color'] = 'transparent'
+    @selectCtx = @selectCanvas.getContext('2d')
+
+  didBecomeActive: (lc) ->
```
Comment (irskep) [inline src/tools/ShapeSelect.coffee]: As the first user of this new tool API who is not me, how easy did you find it?

File: src/core/LiterallyCanvas.coffee (hunk 7 above)
Comment (MandaBrown) [inline src/core/LiterallyCanvas.coffee]: Only because it seemed to me that since we were giving shapes the ability to draw themselves we might as well use it rather than a parallel call to renderShapeInContext. I did that while I was playing with rendering and possibly wanting to change it a bit... 

With this change, the shapes use the canvas renderer but LC doesn't have to know about it any more. It seemed cleaner to me but I'm not overly attached to it if there is a good reason to keep it around in LC.

File: src/tools/ShapeSelect.coffee (hunk 6 above)
Comment (MandaBrown) [inline src/tools/ShapeSelect.coffee]: No worries!

**Comment by irskep** [comment]
Thank you so much for doing this. It's well thought out and follows the project's conventions, as well as providing a feature that people have been asking for since the beginning. You have clearly taken the time to really understand what's going on.

File: src/tools/ShapeSelect.coffee (hunk 9 above)
Comment (MandaBrown) [inline src/tools/ShapeSelect.coffee]: I thought it felt fairly straightforward. The hardest part was looking at examples and seeing a mix and remembering that this is the newer API :)

File: src/tools/ShapeSelect.coffee (hunk 8 above)
Comment (MandaBrown) [inline src/tools/ShapeSelect.coffee]: Yep! That was how I first implemented it. I added methods to the shapes to duplicate themselves with a new color... It was just too fiddly for the app I'm working on. Configurable would be nice.

**Comment by MandaBrown** [comment]
No problem! This is a super fun project to work on, I'm happy when I get a chance to :)

File: src/core/LiterallyCanvas.coffee (hunk 7 above)
Comment (irskep) [inline src/core/LiterallyCanvas.coffee]: Funny thing about that: if you read the top of `src/core/shapes.coffee`, you'll see this in `defineShape`:

```coffee
  Shape.prototype.draw = (ctx, retryCallback) ->
    renderShapeToContext(ctx, this, {retryCallback})
```

It's just a shim to support people who use undocumented behavior. :-) I removed it from the API a couple releases ago.

Comment (MandaBrown) [inline src/core/LiterallyCanvas.coffee]: Yep! That was why I used shape.draw instead of renderShapeToContext! I guess I had the direction of that backwards :)

**Comment by irskep** [comment]
I'm going to try to do a release today BTW (since I finally have time...) which means this probably won't make it in. One of us might have to edit the changelog entry.

File: src/core/LiterallyCanvas.coffee (hunk 7 above)
Comment (MandaBrown) [inline src/core/LiterallyCanvas.coffee]: (Will switch it back; also shape.drawLatest in the shapes in progress loop I assume)

Comment (irskep) [inline src/core/LiterallyCanvas.coffee]: Yep.

**Comment by MandaBrown** [comment]
BTW, @irskep -- one of the reasons we chose this library was because my coworkers and I were very amused by [The New Dork Times](http://steveasleep.com/newdorktimes/).

**Comment by MandaBrown** [comment]
OK, @irskep -- I believe I addressed everything, and I tested with your .json and everything selected OK! (My hex to int was all kinds of wonky -- better now.)

**Comment by gablabelle** [comment]
Hello guys, any idea if and when this PR will be merged? Would love to be able to move shapes also.

**Comment by irskep** [comment]
@MandaBrown merge whenever you like

**Comment by irskep** [comment]
(Sorry for forgetting about this)

**Comment by MandaBrown** [comment]
I forgot about it too! Will look at the conflicts and then merge. Thanks!
